### ISSUE TITLE
Are bindings omnipotent?

### ISSUE THREAD

**Issue by wenxuwan** [dapr/dapr#3722]
I have some doubts about binding. I feel that it is not a very specific abstraction. Many capabilities can be implemented with it. For example, the pub/sub capabilities can also be achieved through outbindinggs/inbindings. So can someone tell me, what is the meaning of bindings? In other words, what types of capabilities should be implemented in bindings。

**Comment by CodeMonkeyLeet** [comment]
@wenxuwan It may be useful to think of Bindings as a lower level abstraction, where the application has more control, but benefits less from the abstractions that Dapr provides. The comparison with PubSub is a useful one as there are integrations like Kafka that have both PubSub and Bindings components:
* Dapr is generally opinionated in the way it provides building blocks like PubSub, which means that it expects components to implement certain specific semantics, such as the competing consumer model for pubsub which the apps can then rely on. The tradeoff in flexibility means that Dapr also can provide additional functionality on top of those components, such as built-in tracing for observability, and Dapr-managed TTL settings per message. 
* On the other hand, Dapr treats Bindings as black boxes, so you don't get tracing IDs and other Dapr features that you would from PubSub. At the same time, the app has to do more work to integrate with a binding but has more flexibility in using the product it is bound with (e.g. non-competing consumer scenarios in Kafka). Bindings allow apps that may not benefit from Dapr's semantics to participate in a daprized microservices deployment.

**Comment by wenxuwan** [comment]
@CodeMonkeyLeet   Thanks for your explanation. As you said, Dapr treats Bindings as black boxes. So user can do anything with it. According to my understanding, dapr treats abstract interfaces as separate implementations, such as pub/sub, but some capabilities that cannot be abstracted can be implemented through bindings. Here binding is just a transparent channel, Is there a problem with my understanding？

**Comment by CodeMonkeyLeet** [comment]
@wenxuwan Yes, insofar as the binding implementation is still restricted to the [Dapr API specification for bindings](https://docs.dapr.io/reference/api/bindings_api/), so for example, you don't have full control over the transport layer still (e.g. Dapr does not natively support streaming channels for bindings).

**Comment by wenxuwan** [comment]
@CodeMonkeyLeet Thanks for your response. In my opinion, an important purpose of Dapr is to abstract a set of standard apis so that applications can realize non-sense migration. So I have some doubts about Binding, because it is not an abstract api, but as you said, the application may not benefit from Dapr's semantics. So compared to using Binding, users who use SDK directly may have better performance, because grpc will have performance loss.

**Comment by wenxuwan** [comment]
> It may be easier to prototype using Dapr's pluggable bindings approach, and as performance tests show issues, swap those out for native integrations and benchmark those.

@CodeMonkeyLeet Thanks again, with your help, the concept and positioning of binding has gradually become clearer
